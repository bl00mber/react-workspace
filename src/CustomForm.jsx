class CustomForm extends React.Component {
  constructor(props) {
    super(props);

    this.selectable = [
      { value: 'grapefruit', text: 'Grapefruit' },
      { value: 'lime', text: 'Lime' },
      { value: 'coconut', text: 'Coconut' },
      { value: 'mango', text: 'Mango' }
    ];

    this.state = {
      textVal: '',
      textareaVal: '',
      selectVal: '',
      selectMultipleVal: [],
    };
  }

  handleChangeText(e) {
    this.setState({ textVal: e.target.value });
  }

  handleChangeTextarea(e) {
    this.setState({ textareaVal: e.target.value });
  }

  handleChangeSelect(e) {
    this.setState({ selectVal: e.target.value });
  }

  handleChangeSelectMultiple(e) {
    const selectedOptions = [...e.target.options]
      .filter((o) => o.selected)
      .map((o) => o.value);
    this.setState({ selectMultipleVal: selectedOptions });
  }

  handleSubmit(e) {
    alert('The form was submitted');
    e.preventDefault();
  }

  render() {
    return (
      <form onSubmit={(e) => this.handleSubmit(e)}>
        <p>
          <label>
            Text:
            <input type="text" value={this.state.textVal} onChange={(e) => this.handleChangeText(e)} />
          </label>
        </p>
        <p>
          <label>
            Textarea:
            <textarea value={this.state.textareaVal} onChange={(e) => this.handleChangeTextarea(e)} />
          </label>
        </p>
        <p>
          <label>
            Pick your favorite La Croix flavor:
            <select value={this.state.selectVal} onChange={(e) => this.handleChangeSelect(e)}>
              {
                this.selectable.map((e) =>
                  <option key={e.value} value={e.value}>{e.text}</option>
                )
              }
            </select>
          </label>
        </p>
        <p>
          <label>
            Pick your favorite La Croix flavors:
            <select multiple={true} value={this.state.selectMultipleVal} onChange={(e) => this.handleChangeSelectMultiple(e)}>
              {
                this.selectable.map((e) =>
                  <option key={e.value} value={e.value}>{e.text}</option>
                )
              }
            </select>
          </label>
          {this.state.selectMultipleVal}
        </p>
        <input type="submit" value="Submit" />
      </form>
    );
  }
}
