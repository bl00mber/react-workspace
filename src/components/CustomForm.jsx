import React from 'react';

class CustomForm extends React.Component {
  static handleSubmit(e) {
    alert('The form was submitted'); // eslint-disable-line no-alert
    e.preventDefault();
  }

  constructor(props) {
    super(props);

    this.selectable = [
      { value: 'grapefruit', text: 'Grapefruit' },
      { value: 'lime', text: 'Lime' },
      { value: 'coconut', text: 'Coconut' },
      { value: 'mango', text: 'Mango' },
    ];

    this.state = {
      textVal: '',
      textareaVal: '',
      selectVal: '',
      selectMultipleVal: [],
    };
  }

  handleChangeText(e) {
    this.setState({ textVal: e.target.value });
  }

  handleChangeTextarea(e) {
    this.setState({ textareaVal: e.target.value });
  }

  handleChangeSelect(e) {
    this.setState({ selectVal: e.target.value });
  }

  handleChangeSelectMultiple(e) {
    const selectedOptions = [...e.target.options]
      .filter(o => o.selected)
      .map(o => o.value);
    this.setState({ selectMultipleVal: selectedOptions });
  }

  render() {
    const {
      textVal,
      textareaVal,
      selectVal,
      selectMultipleVal,
    } = this.state;
    return (
      <form onSubmit={e => CustomForm.handleSubmit(e)}>
        <p>
          <label htmlFor="input01">
            Text:
            <input
              id="input01"
              type="text"
              value={textVal}
              onChange={e => this.handleChangeText(e)}
            />
          </label>
        </p>
        <p>
          <label htmlFor="input02">
            Textarea:
            <textarea
              id="input02"
              value={textareaVal}
              onChange={e => this.handleChangeTextarea(e)}
            />
          </label>
        </p>
        <p>
          <label htmlFor="input03">
            Pick your favorite La Croix flavor:
            <select
              id="input03"
              value={selectVal}
              onChange={e => this.handleChangeSelect(e)}
            >
              {
                this.selectable.map(e => <option key={e.value} value={e.value}>{e.text}</option>)
              }
            </select>
          </label>
        </p>
        <p>
          <label htmlFor="input04">
            Pick your favorite La Croix flavors:
            <select
              id="input04"
              multiple
              value={selectMultipleVal}
              onChange={e => this.handleChangeSelectMultiple(e)}
            >
              {
                this.selectable.map(e => <option key={e.value} value={e.value}>{e.text}</option>)
              }
            </select>
          </label>
          {selectMultipleVal}
        </p>
        <input type="submit" value="Submit" />
      </form>
    );
  }
}

export default CustomForm;
