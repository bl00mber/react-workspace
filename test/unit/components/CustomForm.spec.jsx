import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import CustomForm from '@/components/CustomForm';

describe('CustomForm', () => {
  test('should match enzyme shallow snapshot', () => {
    const wrapper = shallow(<CustomForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should call handleSubmit when submitting form', () => {
    const wrapper = shallow(<CustomForm />);
    jest.spyOn(CustomForm, 'handleSubmit');
    window.alert = jest.fn();
    const e = { preventDefault: jest.fn() };
    wrapper.find('form').simulate('submit', e);
    expect(CustomForm.handleSubmit).toHaveBeenCalledWith(e);
    expect(window.alert).toHaveBeenCalledWith('The form was submitted');
    expect(e.preventDefault).toHaveBeenCalledWith();
  });

  test('should call handleChangeText when changing input01', () => {
    const wrapper = shallow(<CustomForm />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'handleChangeText');
    jest.spyOn(instance, 'setState');
    const e = { target: { value: 'new value' } };
    wrapper.find('input#input01').simulate('change', e);
    expect(instance.handleChangeText).toHaveBeenCalledWith(e);
    expect(instance.setState).toHaveBeenCalledWith({ textVal: 'new value' });
  });

  test('should call handleChangeTextarea when changing input02', () => {
    const wrapper = shallow(<CustomForm />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'handleChangeTextarea');
    jest.spyOn(instance, 'setState');
    const e = { target: { value: 'new value' } };
    wrapper.find('textarea#input02').simulate('change', e);
    expect(instance.handleChangeTextarea).toHaveBeenCalledWith(e);
    expect(instance.setState).toHaveBeenCalledWith({ textareaVal: 'new value' });
  });

  test('should call handleChangeSelect when changing input03', () => {
    const wrapper = shallow(<CustomForm />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'handleChangeSelect');
    jest.spyOn(instance, 'setState');
    const e = { target: { value: 'new value' } };
    wrapper.find('select#input03').simulate('change', e);
    expect(instance.handleChangeSelect).toHaveBeenCalledWith(e);
    expect(instance.setState).toHaveBeenCalledWith({ selectVal: 'new value' });
  });

  test('should call handleChangeSelectMultiple when changing input04', () => {
    const wrapper = shallow(<CustomForm />);
    const instance = wrapper.instance();
    jest.spyOn(instance, 'handleChangeSelectMultiple');
    jest.spyOn(instance, 'setState');
    const e = {
      target: {
        options: [
          { value: 'grapefruit', text: 'Grapefruit' },
          { value: 'lime', text: 'Lime', selected: true },
          { value: 'coconut', text: 'Coconut' },
          { value: 'mango', text: 'Mango' },
        ],
      },
    };
    wrapper.find('select#input04').simulate('change', e);
    expect(instance.handleChangeSelectMultiple).toHaveBeenCalledWith(e);
    expect(instance.setState).toHaveBeenCalledWith({
      selectMultipleVal: ['lime'],
    });
  });
});
