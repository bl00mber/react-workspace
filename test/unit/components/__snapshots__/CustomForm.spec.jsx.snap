// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomForm should match enzyme shallow snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
    <label
      htmlFor="input01"
    >
      Text:
      <input
        id="input01"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </p>
  <p>
    <label
      htmlFor="input02"
    >
      Textarea:
      <textarea
        id="input02"
        onChange={[Function]}
        value=""
      />
    </label>
  </p>
  <p>
    <label
      htmlFor="input03"
    >
      Pick your favorite La Croix flavor:
      <select
        id="input03"
        onChange={[Function]}
        value=""
      >
        <option
          key="grapefruit"
          value="grapefruit"
        >
          Grapefruit
        </option>
        <option
          key="lime"
          value="lime"
        >
          Lime
        </option>
        <option
          key="coconut"
          value="coconut"
        >
          Coconut
        </option>
        <option
          key="mango"
          value="mango"
        >
          Mango
        </option>
      </select>
    </label>
  </p>
  <p>
    <label
      htmlFor="input04"
    >
      Pick your favorite La Croix flavors:
      <select
        id="input04"
        multiple={true}
        onChange={[Function]}
        value={Array []}
      >
        <option
          key="grapefruit"
          value="grapefruit"
        >
          Grapefruit
        </option>
        <option
          key="lime"
          value="lime"
        >
          Lime
        </option>
        <option
          key="coconut"
          value="coconut"
        >
          Coconut
        </option>
        <option
          key="mango"
          value="mango"
        >
          Mango
        </option>
      </select>
    </label>
  </p>
  <input
    type="submit"
    value="Submit"
  />
</form>
`;
